os/arch/arm/src/amebad/amebad_rtc_lowerhalf.c

```diff
@@ -0,0 +1,459 @@
+static int amebad_settime(FAR struct rtc_lowerhalf_s *lower, FAR const struct rtc_time *rtctime);
+
+#ifdef CONFIG_RTC_ALARM
+static int amebad_setalarm(FAR struct rtc_lowerhalf_s *lower, FAR const struct lower_setalarm_s *alarminfo);
+static int amebad_setrelative(FAR struct rtc_lowerhalf_s *lower, FAR const struct lower_setrelative_s *alarminfo);
+static int amebad_cancelalarm(FAR struct rtc_lowerhalf_s *lower, int alarmid);
+#endif
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+/* AMEBAD RTC driver operations */
+
+static const struct rtc_ops_s g_rtc_ops = {
+	.rdtime = amebad_rdtime,
+	.settime = amebad_settime,
+#ifdef CONFIG_RTC_ALARM
+	.setalarm = amebad_setalarm,
+	.setrelative = amebad_setrelative,
+	.cancelalarm = amebad_cancelalarm,
+#endif
+#ifdef CONFIG_RTC_IOCTL
+	.ioctl = NULL,
+#endif
+#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
+	.destroy = NULL,
+#endif
+};
+
+/* AMEBAD RTC device state */
+
+static struct amebad_lowerhalf_s g_rtc_lowerhalf = {
+	.ops = &g_rtc_ops,
+};
+
+alarm_t g_amebad_alarm;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: amebad_alarm_callback
+ *
+ * Description:
+ *   This is the function that is called from the RTC driver when the alarm
+ *   goes off.  It just invokes the upper half drivers callback.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_RTC_ALARM
+alarm_irq_handler amebad_alarm_callback(void)
+{
+	FAR struct amebad_lowerhalf_s *rtc = &g_rtc_lowerhalf;
+
+	/* Sample and clear the callback information to minimize the window in
+	 * time in which race conditions can occur.
+	 */
+	
+	rtc_alarm_callback_t cb = (rtc_alarm_callback_t)rtc->cb;
+	FAR void *arg = (FAR void *)rtc->priv;
+
+	rtc->cb = NULL;
+	rtc->priv = NULL;
+
+	/* Perform the callback */
+
+	if (cb != NULL) {
+		cb(arg, 0);
+	}
+}
+#endif							/* CONFIG_RTC_ALARM */
+
+/****************************************************************************
+ * Name: amebad_rdtime
+ *
+ * Description:
+ *   Implements the rdtime() method of the RTC driver interface
+ *
+ * Input Parameters:
+ *   lower   - A reference to RTC lower half driver state structure
+ *   rcttime - The location in which to return the current RTC time.
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success; a negated errno value is returned
+ *   on any failure.
+ *
+ ****************************************************************************/
+
+static int amebad_rdtime(FAR struct rtc_lowerhalf_s *lower, FAR struct rtc_time *rtctime)
```

Comment:
1. semaphore should be here, I think.
2. This is lowest driver code, so it should not return get_errno. 
    If there was error in somewhere and if errno was set there, here will return that error.
    it should return some error.
Below is my suggestion, please refer to it.
```
int ret;
sem_wait();
timer = up_rtc_time();
ret = gmtime_r(&timer, (FAR struct tm *)rtctime);
sem_post();
if (ret == 0) {
     return -EOVERFLOW;
}
return OK;
```